### PR TITLE
[ISSUE #4817] Add a command to clear broker data from controller for CLI

### LINKED ISSUE #4817: Add a command to clear broker data from controller for CLI
### Add a command to clear broker data from controller for CLI.

**Functional description**
Delete all data on a brokerName or brokerAddress on the Controller for use during O&M 

**e.g**
- Broker shutdown, but Broker data exist on controller
- Some unexpected situation,  Broker data exist on controller but Broker is dead

This command may be used in some cases, etc

### PR DESCRIPTION
**Pull request by mxsm**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

close #4817 

## Brief changelog

- Add a command to clear broker data from controller for CLI

## Verifying this change
as following picture
QQ截图20220814221956
QQ截图20220814223141

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Review by tsunghanjacktsai — COMMENTED** [review]
Hi @mxsm ,

Thx for your contribution. Is there any test for this change?

File: controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java [hunk 1]
```diff
@@ -112,6 +115,24 @@ public void registerBroker(String clusterName, String brokerName, String brokerA
...
+        for (BrokerAddrInfo addrInfo : addrInfoSet) {
+            BrokerLiveInfo brokerLiveInfo = this.brokerLiveTable.get(addrInfo);
+            if (brokerLiveInfo != null && StringUtils.equals(brokerLiveInfo.brokerName, brokerName)) {
+                long last = brokerLiveInfo.lastUpdateTimestamp;
+                long timeoutMillis = brokerLiveInfo.heartbeatTimeoutMillis;
+                return (last + timeoutMillis) >= System.currentTimeMillis();
+            }
+        }
+        return false;
+    }
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java:134]: Is it necessary to add this function? Maybe we can do this in replicasInfoManager

File: controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java [hunk 2]
```diff
@@ -373,6 +374,19 @@ public ControllerResult<Void> getSyncStateData(final List<String> brokerNames) {
         return result;
     }
 
+    public ControllerResult<Void> cleanBrokerData(final CleanControllerBrokerDataRequestHeader requestHeader) {
+        final ControllerResult<Void> result = new ControllerResult<>();
+        String brokerName = requestHeader.getBrokerName();
+        String brokerAddress = requestHeader.getBrokerAddress();
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java:380]: Multiple brokerAddress are not handled here, eg: 192.168.0.1:30911;192.168.0.2:30911

File: controller/src/main/java/org/apache/rocketmq/controller/impl/DLedgerController.java [hunk 3]
```diff
@@ -193,6 +194,32 @@ public RemotingServer getRemotingServer() {
...
+        } else {
+            String[] brokerAddrArray = brokerAddrs.split(";");
+            for (String brokerAddr : brokerAddrArray) {
+                if (brokerAlivePredicate.test(clusterName, brokerAddr)) {
+                    RemotingCommand responseCommand = RemotingCommand.createResponseCommand(ResponseCode.CONTROLLER_INVALID_REQUEST,
+                        String.format("Broker [%s,  %s] is still alive, clean up failure", requestHeader.getBrokerName(), brokerAddr));
+                    return CompletableFuture.completedFuture(responseCommand);
+                }
+            }
+        }
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/DLedgerController.java:217]: How about moving these to replicasInfoManager just like other methods do？

Comment (mxsm) [reply controller/src/main/java/org/apache/rocketmq/controller/impl/DLedgerController.java:217]: I will moving these to replicasInfoManager

File: controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java (hunk 1 above)
Comment (mxsm) [reply controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java:134]: > Is it necessary to add this function? Maybe we can do this in replicasInfoManager

I will do this in replicasInfoManager, and judge by the SyncStateInfo.masterAddress, I will get SyncStateInfo from syncStateSetInfoTable by brokerName. then if SyncStateInfo.masterAddress is not empty, at least one broker with the same BrokerName is alive

**Comment by mxsm** [comment]
> Hi @mxsm ,
> 
> Thx for your contribution. Is there any test for this change?

I will add test for this soon after optimize code logic

**Comment by mxsm** [comment]
Hi @RongtongJin  I has solved conflicts and resubmit code! please help review the code! Thks

**Review by hzh0425 — APPROVED** [review]
Basically LGTM

File: controller/src/main/java/org/apache/rocketmq/controller/Controller.java [hunk 4]
```diff
@@ -106,4 +107,11 @@ CompletableFuture<RemotingCommand> alterSyncStateSet(
      * Get the remotingServer used by the controller, the upper layer will reuse this remotingServer.
      */
     RemotingServer getRemotingServer();
+
+    /**
+     * Clean controller broker data, the broker which  shut down or offline
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/Controller.java:112]: “Clean controller broker data, the broker which  shut down or offline” -> "Clean controller broker data"

File: controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java [hunk 5]
```diff
@@ -387,6 +429,29 @@ private void handleElectMaster(final ElectMasterEvent event) {
         }
     }
 
+    private void handleCleanBrokerDataEvent(final CleanBrokerDataEvent event) {
+        final String brokerName = event.getBrokerName();
+        final Set<String> brokerAddressSet = event.getBrokerAddressSet();
+
+        if (!isContainsBroker(brokerName) || null == brokerAddressSet || brokerAddressSet.isEmpty()) {
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java:436]: Why do we need `this.replicaInfoTable.remove(brokerName)` and `this.syncStateSetInfoTable.remove(brokerName)` when !isContainsBroker(brokerName)?

File: controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java [hunk 6]
```diff
@@ -316,6 +320,41 @@ public ControllerResult<Void> getSyncStateData(final List<String> brokerNames) {
...
+        if (StringUtils.isNotBlank(brokerAddrs)) {
+            brokerAddressSet = Stream.of(brokerAddrs.split(";")).collect(Collectors.toSet());
+            for (String brokerAddr : brokerAddressSet) {
+                if (brokerAlivePredicate.test(clusterName, brokerAddr)) {
+                    String remark = String.format("Broker [%s,  %s] is still alive, clean up failure", requestHeader.getBrokerName(), brokerAddr);
+                    result.setCodeAndRemark(ResponseCode.CONTROLLER_INVALID_REQUEST, remark);
+                    return result;
+                }
+            }
+        }
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java:348]: Hi, @mxsm I don't think it is necessary to check whether the brokers are alive before cleaning. Even if the living brokers are cleaned, they will register again.

File: tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        opt = new Option("b", "brokerAddress", true, "The address of the broker which requires to clean metadata. eg: 192.168.0.1:30911;192.168.0.2:30911");
+        opt.setRequired(false);
+        options.addOption(opt);
+
+        opt = new Option("n", "brokerName", true, "The broker name of the replicas that require to be manipulated");
+        opt.setRequired(true);
+        options.addOption(opt);
+
+        opt = new Option("c", "clusterName", true, "the clusterName of broker");
```
Comment (RongtongJin) [inline tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:55]: Hi, @mxsm Could you explain the meaning of clusterName here？

Comment (mxsm) [reply tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:55]: > Hi, @mxsm Could you explain the meaning of clusterName here？

In Controller server ReplicasInfoManager#cleanBrokerData I need it and brokerAddrs to test broker is active.
```java
# ReplicasInfoManager#cleanBrokerData
    public ControllerResult<Void> cleanBrokerData(final CleanControllerBrokerDataRequestHeader requestHeader,
        final BiPredicate<String, String> brokerAlivePredicate) {
        final ControllerResult<Void> result = new ControllerResult<>();

        final String clusterName = requestHeader.getClusterName();
        final String brokerName = requestHeader.getBrokerName();
        final String brokerAddrs = requestHeader.getBrokerAddress();

        //if SyncStateInfo.masterAddress is not empty, at least one broker with the same BrokerName is alive
        SyncStateInfo syncStateInfo = this.syncStateSetInfoTable.get(brokerName);
        if (StringUtils.isBlank(brokerAddrs) && null != syncStateInfo && StringUtils.isNotEmpty(syncStateInfo.getMasterAddress())) {
            String remark = String.format("Broker %s is still alive, clean up failure", requestHeader.getBrokerName());
            result.setCodeAndRemark(ResponseCode.CONTROLLER_INVALID_REQUEST, remark);
            return result;
        }
        Set<String> brokerAddressSet = null;
        if (StringUtils.isNotBlank(brokerAddrs)) {
            brokerAddressSet = Stream.of(brokerAddrs.split(";")).collect(Collectors.toSet());
            for (String brokerAddr : brokerAddressSet) {
                if (brokerAlivePredicate.test(clusterName, brokerAddr)) {
                    String remark = String.format("Broker [%s,  %s] is still alive, clean up failure", requestHeader.getBrokerName(), brokerAddr);
                    result.setCodeAndRemark(ResponseCode.CONTROLLER_INVALID_REQUEST, remark);
                    return result;
                }
            }
        }
        if (isContainsBroker(brokerName)) {
            final CleanBrokerDataEvent event = new CleanBrokerDataEvent(brokerName, brokerAddressSet);
            result.addEvent(event);
            return result;
        }
        result.setCodeAndRemark(ResponseCode.CONTROLLER_INVALID_REQUEST, String.format("Broker %s is not existed", brokerName));
        return result;
    }
```

Comment (RongtongJin) [reply tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:55]: OK, I got it. But we need to discuss whether we can only clean up non-living brokers.

Comment (mxsm) [reply tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:55]: > OK, I got it. But we need to discuss whether we can only clean up non-living brokers.

Can we give the user an option to decide whether only clean up non-living brokers.

File: controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java (hunk 6 above)
Comment (mxsm) [reply controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java:348]: I got it

File: tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java (hunk 7 above)
Comment (RongtongJin) [reply tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:55]: > > OK, I got it. But we need to discuss whether we can only clean up non-living brokers.
> 
> @RongtongJin Can we give the user an option to decide whether only clean up non-living brokers.

Good suggestion

Comment (mxsm) [reply tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:55]: > > > OK, I got it. But we need to discuss whether we can only clean up non-living brokers.
> > 
> > 
> > @RongtongJin Can we give the user an option to decide whether only clean up non-living brokers.
> 
> Good suggestion

I will follow this current idea to implement @RongtongJin

File: tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        boolean isCleanLivingBroker = false;
+        if (commandLine.hasOption('c')) {
+            clusterName = commandLine.getOptionValue('c').trim();
+        }
+        if (commandLine.hasOption('b')) {
+            brokerAddress = commandLine.getOptionValue('b').trim();
+        }
+        if (commandLine.hasOption('l')) {
+            isCleanLivingBroker = true;
+        }
```
Comment (RongtongJin) [inline tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:86]: The specific value of commandLine shall prevail, and cannot be directly set to true

Comment (mxsm) [reply tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:86]: OK I got it, need user to set the value

File: tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        opt = new Option("n", "brokerName", true, "The broker name of the replicas that require to be manipulated");
+        opt.setRequired(true);
+        options.addOption(opt);
+
+        opt = new Option("c", "clusterName", true, "the clusterName of broker");
+        opt.setRequired(false);
+        options.addOption(opt);
+
+        opt = new Option("l", "cleanLivingBroker", true, " whether clean up living brokers. value is false or true");
+        opt.setRequired(true);
```
Comment (RongtongJin) [inline tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java:61]: IMO, cleanLivingBroker option is not required, the default value is false.

**Comment by RongtongJin** [comment]
Hi, @mxsm Could you rebase or merge the latest develop branch to this pull request？ Develop branch add new required status checks, this pull request must pass the checker before merging.

**Review by RongtongJin — APPROVED** [review]

**Pull request merged**